File: android/app/src/main/java/io/github/magnusencoded/setlist2spotify/data/Filing.kt
package io.github.magnusencoded.setlist2spotify.data

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.widget.Toast
import io.github.magnusencoded.setlist2spotify.data.setlistfm.FmSetlist

/**
 * The night's facts, parked in the notification shade while you fill in setlist.fm's
 * form in the browser.
 *
 * **The problem this solves.** setlist.fm's add form takes no prefill parameters
 * (verified, see [SETLISTFM_ADD_URL]), so the clipboard is the only channel into it —
 * and the clipboard holds exactly one thing while the form asks for five. Worse, the
 * app screen that knows the artist, venue, town and date is by definition *not* on
 * screen once Chrome has the foreground. So those four were crossing the app switch
 * in the Historian's memory, which is where a misremembered venue on a shared public
 * record comes from.
 *
 * **Why a notification.** It is the one surface Android keeps in reach of another
 * app. One notification per field, grouped: pull the shade down over the form, tap
 * the value you need, paste, carry on. Any order, any number of times — an
 * autocomplete that swallowed the first paste is answered by tapping again rather
 * than by going back to the app.
 *
 * A tap must **not** open the app. Foregrounding would throw you out of the browser
 * mid-form, which is the whole thing being avoided, so the tap is a broadcast to
 * [CopyReceiver] and the app never comes forward.
 *
 * ponytail: no re-posting, no stepper, no state. Each notification is a fixed value
 * with a fixed action, and swiping the group summary clears all of them.
 */
private const val FILING_CHANNEL = "filing"
private const val FILING_GROUP = "io.github.magnusencoded.setlist2spotify.FILING"

/** Summary owns the base id; each field sits above it. */
private const val FILING_SUMMARY_ID = 8100

const val EXTRA_FILING_LABEL = "filing_label"
const val EXTRA_FILING_VALUE = "filing_value"

/**
 * Puts one field on the clipboard, from the shade, without waking the app.
 *
 * **The bit that has to be true for any of this to work:** a background broadcast can
 * still *write* the clipboard. Android 10 restricted clipboard access for apps without
 * focus — that restriction is on reading, and writing is what this does. If a future
 * Android closes that too, this fails in the worst possible way: silently, leaving
 * whatever was on the clipboard before to be pasted into a public database. Hence the
 * toast — it is not a courtesy, it is the only evidence the copy happened, and its
 * absence is the signal that this approach has died.
 */
class CopyReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        val label = intent.getStringExtra(EXTRA_FILING_LABEL) ?: return
        val value = intent.getStringExtra(EXTRA_FILING_VALUE) ?: return
        context.getSystemService(ClipboardManager::class.java)
            ?.setPrimaryClip(ClipData.newPlainText(label, value))
        Toast.makeText(context, "$label copied — paste it in", Toast.LENGTH_SHORT).show()
    }
}

/**
 * Post one notification per field of [setlist]'s filing, plus the summary that groups
 * them. Replaces any previous filing: ids are positional, so filing a second night
 * overwrites the first rather than stacking two sets of values in the shade — two
 * "Venue" notifications from different gigs is exactly the confusion this exists to
 * prevent.
 *
 * Silently does nothing when there is nothing to file. Notification permission is the
 * caller's problem: [postFiling] is a best-effort *upgrade* to the handoff, and the
 * clipboard-and-browser path it accompanies must work whether or not this ran.
 */
fun postFiling(context: Context, setlist: FmSetlist, log: StoredLog) {
    val fields = filingFields(setlist, log)
    if (fields.isEmpty()) return
    val manager = context.getSystemService(NotificationManager::class.java) ?: return
    manager.createNotificationChannel(
        NotificationChannel(
            FILING_CHANNEL,
            "Filing a night",
            // Low: this is a tray of values you went looking for, not news. It must
            // never buzz — you are standing in a form when it appears.
            NotificationManager.IMPORTANCE_LOW,
        ).apply { description = "The night's details, ready to paste into setlist.fm." },
    )
    clearFiling(context)
    fields.forEachIndexed { i, field ->
        val intent = Intent(context, CopyReceiver::class.java)
            .putExtra(EXTRA_FILING_LABEL, field.label)
            .putExtra(EXTRA_FILING_VALUE, field.value)
        val tap = PendingIntent.getBroadcast(
            context,
            FILING_SUMMARY_ID + 1 + i,
            intent,
            // IMMUTABLE is required from 31 and correct anyway — nothing outside this
            // app has any business editing which value a tap copies. UPDATE_CURRENT so
            // re-filing the same slot carries the new night's value, not the old one.
            PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT,
        )
        manager.notify(
            FILING_SUMMARY_ID + 1 + i,
            Notification.Builder(context, FILING_CHANNEL)
                .setSmallIcon(android.R.drawable.ic_menu_edit)
                .setContentTitle(field.label)
                .setContentText(field.shown)
                // The songs are a paste, not a line: expanded, the shade shows what is
                // actually going to land in the form.
                .setStyle(Notification.BigTextStyle().bigText(field.value))
                .setContentIntent(tap)
                // A tap is a copy, not a dismissal. Pasting the venue twice because the
                // autocomplete ate the first one is the ordinary case, not an error.
                .setAutoCancel(false)
                .setGroup(FILING_GROUP)
                .build(),
        )
    }
    manager.notify(
        FILING_SUMMARY_ID,
        Notification.Builder(context, FILING_CHANNEL)
            .setSmallIcon(android.R.drawable.ic_menu_edit)
            .setContentTitle("Filing ${setlist.artist?.name ?: "this night"}")
            .setContentText("Tap a value to copy it, then paste. Swipe to finish.")
            .setGroup(FILING_GROUP)
            .setGroupSummary(true)
            // The summary is the "done" gesture: dismissing it takes the whole tray
            // with it, so finishing costs one swipe rather than five.
            .build(),
    )
}

/** Take the tray down. Filing again, or finishing, both end here. */
fun clearFiling(context: Context) {
    val manager = context.getSystemService(NotificationManager::class.java) ?: return
    // One more than any filing can have, so a night with fewer fields than the last
    // one leaves nothing of the last one behind.
    for (id in FILING_SUMMARY_ID..FILING_SUMMARY_ID + 8) manager.cancel(id)
}

File: ios/StationToStation/Data/Filing.swift
import Foundation

// Filing a night with setlist.fm: the values the form asks for, and the door to it.
//
// The Swift twin of Android's `setlistPaste`, `filingFields` and `setlistEditEntry`
// (`data/Bill.kt`), asserted by the same cases. Pure — nothing here copies, opens or
// announces anything; what carries these across the app switch is the caller's, and
// the two platforms have different surfaces for that.

let setlistfmAddURL = "https://www.setlist.fm/edit"

/// The **Log** as setlist.fm's Text Field editor wants it: bare titles, one per line,
/// in the order they were played.
///
/// A Gap pastes as setlist.fm's own unknown marker rather than as nothing — dropping
/// it would publish a set silently claiming that song was not played.
func setlistPaste(_ log: StoredLog) -> String {
    log.songs
        .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
        .map { $0.isEmpty ? "@Unknown[]" : $0 }
        .joined(separator: "\n")
}

/// One value the setlist.fm form wants, ready to be handed over.
///
/// `shown` and `value` differ for exactly one field — the songs, where the value is a
/// fourteen-line paste and what you want to *read* is "14 songs, in order". Everywhere
/// else they are the same string.
struct FilingField: Equatable {
    let label: String
    let shown: String
    let value: String
}

/// Everything the setlist.fm add form asks for, in the order it asks for it.
///
/// The clipboard holds one thing. The form wants five, and the app screen that knows
/// them is not on screen once the browser is — so the night's facts were being carried
/// across the app switch in the Historian's head, which is where a wrong venue comes
/// from.
///
/// **The order is the form's:** Add artist, Select event date, Add venue, then the
/// songs on the step after. Date before venue is not a preference — the venue field is
/// *disabled* until a date is set. A tray that offered Venue first would be offering a
/// value with nowhere to go.
///
/// Blank fields are dropped rather than posted empty: a night with no town typed in
/// should offer four values, not four and a lie.
func filingFields(_ setlist: FmSetlist, _ log: StoredLog) -> [FilingField] {
    var out: [FilingField] = []
    if let artist = setlist.artist?.name.nilIfBlank {
        out.append(FilingField(label: "Artist", shown: artist, value: artist))
    }
    // Shown as "5 August 2026" rather than 05-08-2026, because this one is **picked,
    // not pasted**: the field opens a calendar widget, so no string can land in it.
    // What the Historian actually does with this value is find that day in a month
    // grid, and a written-out month is the form of it that matches the gesture. The
    // raw date stays the copied value; it costs nothing and the clipboard is the
    // wrong place to editorialise.
    if let date = setlist.eventDate?.nilIfBlank {
        out.append(FilingField(label: "Date", shown: setlist.readableDate() ?? date, value: date))
    }
    if let venue = setlist.venue?.name?.nilIfBlank {
        out.append(FilingField(label: "Venue", shown: venue, value: venue))
    }
    if let city = setlist.venue?.city?.name?.nilIfBlank {
        out.append(FilingField(label: "City", shown: city, value: city))
    }
    if !log.songs.isEmpty {
        // Gaps are counted in, because they are in the paste and will appear in the
        // form. "13 songs" beside a fourteen-line paste is the kind of small
        // disagreement that makes someone distrust the whole handoff.
        let unnamed = log.gaps > 0 ? " \u{00B7} \(log.gaps) unnamed" : ""
        out.append(FilingField(label: "Songs",
                               shown: "\(log.songs.count) songs, in order" + unnamed,
                               value: setlistPaste(log)))
    }
    return out
}

/// Where the Historian is sent to file this night, and it is one of two places.
///
/// A gig setlist.fm already has — including the empty-setlist case, a record with a
/// page and no songs — goes to *its own page*, one click from the right edit form. A
/// gig they have never heard of goes to the generic add flow, which is the only door
/// there is. The clipboard carries the set either way.
func setlistEditEntry(_ setlist: FmSetlist) -> String { setlist.url ?? setlistfmAddURL }
